components/filter/filter.js

```diff
@@ -0,0 +1,270 @@
+							@d2l-filter-dimension-slot-change="${this._handleSlotChange}"
+							@slotchange="${this._handleSlotChange}"
+						></slot>
+						${dimensions}
+					</d2l-menu>
+				</d2l-dropdown-menu>
+			</d2l-dropdown-button-subtle>
+		`;
+	}
+
+	_buildDimension(dimension) {
+		let dimensionHTML;
+		switch (dimension.type) {
+			case 'd2l-filter-dimension-set':
+				dimensionHTML = this._createSetDimension(dimension);
+				break;
+		}
+		return html`
+			<d2l-hierarchical-view
+				@d2l-hierarchical-view-show-complete="${this._handleDimensionShowComplete}"
+				@d2l-hierarchical-view-show-start="${this._handleDimensionShowStart}"
+				id="${dimension.key}">
+
+				${dimensionHTML}
+			</d2l-hierarchical-view>
+		`;
+	}
+
+	_buildDimensions() {
+		return this._dimensions.map((dimension) => {
+			const builtDimension = this._buildDimension(dimension);
+			return html`<d2l-menu-item text="${dimension.text}">
+				${builtDimension}
+			</d2l-menu-item>`;
+		});
+	}
+
+	_buildHeader() {
+		if (!this._activeDimensionKey) return null;
+
+		const dimensionText = this._dimensions.find(dimension => dimension.key === this._activeDimensionKey).text;
+		return html`
+			<div slot="header">
+				<div class="d2l-filter-dimension-header">
+					<d2l-button-icon
+						@click="${this._handleDimensionHide}"
+						icon="tier1:chevron-left"
+						text="${this.localize('components.menu-item-return.returnCurrentlyShowing', 'menuTitle', dimensionText)}">
+					</d2l-button-icon>
+					<div class="d2l-filter-dimension-header-text d2l-body-standard">${dimensionText}</div>
+				</div>
+			</div>
+		`;
+	}
+
+	_createSetDimension(dimension) {
+		return html`
+			<d2l-list
+				@d2l-list-selection-change="${this._handleChangeSetDimension}"
+				extend-separators>
+
+				${dimension.values.map(item => html`
+					<d2l-list-item
+						key="${item.key}"
+						selectable
+						?selected="${item.selected}"
+						slim>
+
+						<div class="d2l-filter-dimension-set-value-text">${item.text}</div>
+					</d2l-list-item>
+				`)}
+			</d2l-list>
+		`;
+	}
+
+	_dispatchChangeEvent(eventDetail) {
+		this.dispatchEvent(new CustomEvent('d2l-filter-change', { bubbles: false, composed: false, detail: eventDetail }));
+	}
+
+	_getSlottedNodes(slot) {
+		const nodes = slot.assignedNodes();
+		return nodes.filter((node) => node.nodeType === Node.ELEMENT_NODE && node.tagName.toLowerCase() !== 'template');
+	}
+
+	_handleChangeSetDimension(e) {
+		const dimensionKey = e.composedPath()[0].parentNode.id;
+		const valueKey = e.detail.key;
+		const selected = e.detail.selected;
+
+		// Update the corresponding d2l-filter-dimension-set-value to keep them in sync
+		const dimension = this.querySelector(`d2l-filter-dimension-set[key="${dimensionKey}"`);
+		const slot = dimension.shadowRoot.querySelector('slot');
+		const items = this._getSlottedNodes(slot);
+		const item = items.find(item => item.key === valueKey);
+		item.selected = selected;
```

Comment:
This still doesn't quite feel right. Like I get why it's necessary... querying into the DOM and setting properties on the dimension values to reflect the menu's state so that if the consumer reads those properties they're accurate... it just doesn't feel right.

I'm wondering how necessary it is? Like maybe we could say they're only used to build up the initial state, and then after that consumers need to use the `@change` event on the filter to react to changes? That would be similar to how `<input type="checkbox" checked>` is only correct initially and after that you use the value. I dunno, just seems like a lot of work.

Reply:
So _maybe_ we can get away with that... basically, the user would select an option, the filter fires an event, and the consumer needs to then update the data it uses to render the `?select=<whatever>` portion if it's going to need to care about that value after the initial render.  The main places I see this being important are the HM filter framework (where the filter accepts an href that understands the filters applied and updates the UI component accordingly, every time a filter is applied) and with the filter tags (where they can remove a filter tag, and the corresponding value is unselected in the filter component.

But I _think_ they would naturally keep their own data up-to-date anyways, so maybe I am pre-fixing a problem we may not see.  I could remove it for now and see if it comes up replacing the HM filter or writing the filter tags

Reply:
I think they're going to need to keep internal state anyway, since the next time their component re-renders they'll also re-render the filter and blow away anything that changed anyway right? So us updating that state ourselves isn't really helping them if they're just going to overwrite it the next time their component renders.

Reply:
Hmm... unless Lit gets confused and thinks things have or haven't changed when they re-render -- which was maybe the original motivation for doing this.

Reply:
Yeah, my worry was that they would try to set something to the same value and Lit wouldn't recognize anything as different and `updated` would never get called.  _But in theory_, if they kept things up-to-date every time, they wouldn't ever get into a state where things are off.  I was trying to foolproof things for them, but I can't remember if I actually _saw_ this as a problem, or figured it would be

Reply:
I'm not sure either what's needed. Might be worth doing a quick experiment to see if keeping these properties up-to-date ourselves helps or hinders.

Reply:
Looked at two current but different HM usages (the shared HM filter and My Courses) and both are already keeping their own data up-to-date and updating the component upon receiving the event anyways.  Removing this until we see a use case, but I think you're right - it's not hard for the consumer to do this if they need to, and it seems to fit the natural mental model.